File: src/vidoc/model/Vidoc.ts
import { AuthorInformation } from "./AuthorInformation";
import {
    EditorPosition,
    EditorSelection,
    FocusInformation,
} from "./FocusInformation";

export interface VidocMetadata {
    createdAt: string;
    createdBy: AuthorInformation;
    focusInformation: FocusInformation;
}

export interface BoundsInformation {
    x: number;
    y: number;
    width: number;
    height: number;
}

export interface ScreenInformation extends BoundsInformation {
    index: number;
    scale: {
        x: number;
        y: number;
    };
}

export interface RecordingAreaInformation {
    bounds: BoundsInformation;
    screen: ScreenInformation;
}

export interface Vidoc {
    id: string;
    metadata: VidocMetadata;
    tmpVideoFilePaths: string[]; //relative
    mergedTmpVideoFilePath?: string; //relative
    relativeFilePathMetadata: string;
    speechToText?: SpeechToTextInformation;
    recordingAreaInformation?: RecordingAreaInformation;
}

export interface SpeechToTextInformation {
    text: string;
    payload: any;
}

export interface LocalMetadataLocalVideoVidoc extends Vidoc {
    relativeFilePathToVideo: string;
}

export interface LocalMetaDataRemoteVideoVidoc extends Vidoc {
    remoteVideoUrl: string;
    uploadUrl?: string;
}

export interface PositionedVidocInstance {
    range: EditorSelection;
    vidoc: Vidoc;
}

export interface VidocReference {
    relativeFilePath: string;
    vidoc: Vidoc;
    editorPosition: EditorPosition;
}

export interface HTMLPageOutput {
    html: string;
    path: string;
}

File: src/intellij/vidoc/src/main/kotlin/com/hmdevconsulting/vidoc/model/Vidoc.kt
package com.hmdevconsulting.vidoc.model


data class VidocMetadata(
    val createdAt: String,
    val createdBy: AuthorInformation,
    val focusInformation: FocusInformation
)

data class Vidoc(
    val id: String,
    val metadata: VidocMetadata,
    val tmpVideoFilePath: String,
    val relativeFilePathMetadata: String,
    val speechToText: SpeechToTextInformation? = null,
    val relativeFilePathToVideo: String? = null,
    val remoteVideoUrl: String? = null
)

data class SpeechToTextInformation(
    val text: String,
    val payload: Any
)

data class PositionedVidocInstance(
    val range: EditorSelection,
    val vidoc: Vidoc
)

data class VidocReference(
    val relativeFilePath: String,
    val vidoc: Vidoc,
    val editorPosition: EditorPosition
)

data class EditorPosition(
    val lineIndex: Int,
    val charIndex: Int,
    val lineContent: String
)

data class EditorSelection(
    val from: EditorPosition,
    val to: EditorPosition,
    val text: String
)

data class GivenFocusInformation(
    val cursorPosition: EditorPosition,
    val currentSelection: EditorSelection,
    val currentlyOpenedFileRelativeFilePath: String
)

typealias FocusInformation = GivenFocusInformation?

data class GivenAuthorInformation(
    val username: String? = null,
    val fullName: String? = null,
    val email: String? = null
)

typealias AuthorInformation = GivenAuthorInformation?
